File: app/src/main/java/dev/patrickgold/florisboard/ime/nlp/StagedSuggestionList.kt
/*
 * Copyright (C) 2021 Patrick Goldinger
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package dev.patrickgold.florisboard.ime.nlp

class StagedSuggestionList<T : Any, F : Comparable<F>>(
    private val maxSize: Int
) : Collection<WeightedToken<T, F>> {
    private val internalArray: Array<WeightedToken<T, F>?> = Array(maxSize) { null }
    private var internalSize: Int = 0

    override val size: Int
        get() = internalSize

    fun add(token: T, freq: F): Boolean {
        if (internalSize < maxSize) {
            internalArray[internalSize++] = WeightedToken(token, freq)
            internalArray.sortByDescending { it?.freq }
            return true
        } else {
            if (internalArray.last()!!.freq < freq) {
                internalArray[internalArray.lastIndex] = WeightedToken(token, freq)
                internalArray.sortByDescending { it?.freq }
                return true
            }
            return false
        }
    }

    fun canAdd(freq: F): Boolean {
        return internalSize < maxSize || internalArray.last()!!.freq < freq
    }

    fun clear() {
        for (n in internalArray.indices) {
            internalArray[n] = null
        }
        internalSize = 0
    }

    override fun contains(element: WeightedToken<T, F>): Boolean = internalArray.contains(element)

    override fun containsAll(elements: Collection<WeightedToken<T, F>>): Boolean {
        elements.forEach { if (!contains(it)) return false }
        return true
    }

    @Throws(IndexOutOfBoundsException::class)
    operator fun get(index: Int): WeightedToken<T, F> {
        val element = getOrNull(index)
        if (element == null) {
            throw IndexOutOfBoundsException("The specified index $index is not within the bounds of this list!")
        } else {
            return element
        }
    }

    fun getOrNull(index: Int): WeightedToken<T, F>? {
        return internalArray.getOrNull(index)
    }

    override fun isEmpty(): Boolean = internalSize <= 0

    override fun iterator(): Iterator<WeightedToken<T, F>> {
        return StagedIterator(this)
    }

    class StagedIterator<T : Any, F : Comparable<F>> internal constructor (
        private val stagedSuggestionList: StagedSuggestionList<T, F>
    ) : Iterator<WeightedToken<T, F>> {
        var index = 0

        override fun next(): WeightedToken<T, F> = stagedSuggestionList[index++]

        override fun hasNext(): Boolean = stagedSuggestionList.getOrNull(index) != null
    }
}

File: app/src/main/cpp/ime/nlp/staged_suggestion_list.cpp
/*
 * Copyright (C) 2021 Patrick Goldinger
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "staged_suggestion_list.h"

#include <utility>

using namespace ime::nlp;

SuggestionList::SuggestionList(size_t _maxSize) :
    maxSize(_maxSize), internalSize(0), internalArray(new WeightedToken*[_maxSize]), isPrimaryTokenAutoInsert(false)
{
    // Initialize the internal array to null pointers
    for (size_t n = 0; n < maxSize; n++) {
        internalArray[n] = nullptr;
    }
}

SuggestionList::~SuggestionList() {
    delete[] internalArray;
}

bool SuggestionList::add(word_t word, freq_t freq) {
    auto entryIndex = indexOfWord(word);
    if (entryIndex.has_value()) {
        // Word exists already
        auto entry = get(entryIndex.value());
        if (entry->freq < freq) {
            // Need to update freq
            entry->freq = freq;
        } else {
            return false;
        }
    } else {
        if (internalSize < maxSize) {
            internalArray[internalSize++] = new WeightedToken(std::move(word), freq);
        } else {
            WeightedToken *last = internalArray[internalSize - 1];
            if (last->freq < freq) {
                internalArray[internalSize - 1] = new WeightedToken(std::move(word), freq);
            } else {
                return false;
            }
        }
    }
    std::sort(internalArray, internalArray + internalSize, std::greater<>());
    return true;
}

void SuggestionList::clear() {
    for (size_t n = 0; n < internalSize; n++) {
        delete internalArray[n];
        internalArray[n] = nullptr;
    }
    internalSize = 0;
    isPrimaryTokenAutoInsert = false;
}

bool SuggestionList::contains(WeightedToken &element) {
    return indexOf(element).has_value();
}

bool SuggestionList::containsWord(const word_t &word) {
    return indexOfWord(word).has_value();
}

WeightedToken *SuggestionList::get(size_t index) {
    if (index < 0 || index >= maxSize) return nullptr;
    return internalArray[index];
}

std::optional<size_t> SuggestionList::indexOf(WeightedToken &element) {
    for (size_t n = 0; n < internalSize; n++) {
        if (element == *internalArray[n]) {
            return n;
        }
    }
    return std::nullopt;
}

std::optional<size_t> SuggestionList::indexOfWord(const word_t &word) {
    for (size_t n = 0; n < internalSize; n++) {
        if (word == internalArray[n]->data) {
            return n;
        }
    }
    return std::nullopt;
}

bool SuggestionList::isEmpty() const {
    return internalSize == 0;
}

size_t SuggestionList::size() const {
    return internalSize;
}
